### PR TITLE
Implement `AdsClient`/`Ads` observers

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/26592

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_ads/browser/ads_service.h [hunk 1]
```diff
@@ -87,6 +97,8 @@ class AdsService : public KeyedService {
   // Called if a browser upgrade is required to serve ads.
   virtual bool NeedsBrowserUpgradeToServeAds() const = 0;
 
+  virtual void NotifyRewardsWalletVerified() const = 0;
```
Comment (tmancey) [inline components/brave_ads/browser/ads_service.h:100]: `NotifyUserDidVerifyRewardsWallet`

**Comment by tmancey** [comment]
The build is failing due to some missing files. Thanks

File: vendor/bat-native-ads/include/bat/ads/ads_client.h [hunk 2]
```diff
@@ -26,6 +28,9 @@ class ADS_EXPORT AdsClient {
  public:
   virtual ~AdsClient() = default;
 
+  virtual void AddBatAdsClientObserver(
```
Comment (tmancey) [inline vendor/bat-native-ads/include/bat/ads/ads_client.h:31]: Can we please add a comment, i.e. `// Called to add an ads client observer.`

File: components/brave_ads/browser/ads_service.h (hunk 1 above)
Comment (aseren) [reply components/brave_ads/browser/ads_service.h]: fixed

File: components/brave_ads/browser/ads_service.h [hunk 3]
```diff
@@ -87,6 +97,8 @@ class AdsService : public KeyedService {
   // Called if a browser upgrade is required to serve ads.
   virtual bool NeedsBrowserUpgradeToServeAds() const = 0;
 
+  virtual void NotifyUserDidVerifyRewardsWallet() const = 0;
```
Comment (tmancey) [inline components/brave_ads/browser/ads_service.h:100]: `// Called to notify that a user did verify their Rewards wallet.`

File: ios/browser/api/ads/brave_ads.mm [hunk 4]
```diff
@@ -473,18 +482,29 @@ - (void)removeAllHistory:(void (^)(BOOL))completion {
...
+- (void)addObserver:(ads::AdsClientObserver*)observer {
+  ads->AddObserver(observer);
+}
+
+- (void)removeObserver:(ads::AdsClientObserver*)observer {
+  ads->RemoveObserver(observer);
+}
+
+- (void)notifyDidInitializeAds {
+}
```
Comment (kylehickinson) [inline ios/browser/api/ads/brave_ads.mm:494]: 1. These aren't used anywhere?
2. These aren't in the headers
3. We can't use it directly like this, we would need a wrapper around `ads::AdsClientObserver` in order to expose it in the headers/to brave-ios

File: ios/browser/api/ads/brave_ads.mm [hunk 5]
```diff
@@ -268,6 +271,7 @@ - (void)initialize:(void (^)(bool))completion {
   adEventHistory = new ads::AdEventHistory();
 
   adsClient = new AdsClientIOS(self);
+  adsClientObserverNotifier = new AdsClientObserverNotifier();
```
Comment (kylehickinson) [inline ios/browser/api/ads/brave_ads.mm:274]: What does this notifier do? I don't see any connection to the original `ads` class which had these methods before. Is there a binding/init of some sort that we're missing?

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: ios/browser/api/ads/brave_ads.mm (hunk 4 above)
Comment (tmancey) [reply ios/browser/api/ads/brave_ads.mm]: WIP. I did not mean to open the PR for review :-)

File: ios/browser/api/ads/brave_ads.mm (hunk 5 above)
Comment (tmancey) [reply ios/browser/api/ads/brave_ads.mm]: Ditto

**Pull request closed without merging**